File: ml-parcial-IIND3118/ml-parcial-IIND3118/temas/A07_regresion_lineal/python/A07_regresion_lineal.py
"""
A7. Regresión lineal
Tema A07_regresion_lineal — Introducción al Machine Learning (IIND3118)

CÓMO USARLO
  1. Cambia el bloque marcado con ✏️ abajo por tu dataset.
  2. Corre el archivo:  python A07_regresion_lineal.py
  3. Las gráficas se guardan en la carpeta figuras/ de este tema.

La explicación del método está en el README.md de esta misma carpeta.
"""

import os, sys, warnings
warnings.filterwarnings("ignore")

import numpy as np
import pandas as pd
import matplotlib
matplotlib.use("Agg")          # no abre ventanas: funciona en cualquier entorno
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split, KFold, cross_val_score, GridSearchCV
from sklearn.preprocessing import scale
from sklearn.linear_model import (LinearRegression, Ridge, Lasso, ElasticNet,
                                  RidgeCV, LassoCV, ElasticNetCV, LogisticRegression)
from sklearn.neighbors import KNeighborsRegressor, KNeighborsClassifier
from sklearn.metrics import (mean_squared_error, mean_absolute_error, r2_score,
                             accuracy_score, confusion_matrix, classification_report,
                             roc_curve, roc_auc_score)
from sklearn.decomposition import PCA, KernelPCA
from sklearn.cluster import KMeans, DBSCAN
from sklearn.ensemble import IsolationForest
from sklearn.neighbors import LocalOutlierFactor
from sklearn import manifold

# ── Guardado automático de figuras ──────────────────────────────────────
AQUI = os.path.dirname(os.path.abspath(__file__))
FIGS = os.path.join(AQUI, "..", "figuras")
os.makedirs(FIGS, exist_ok=True)
_n = [0]
def _save():
    _n[0] += 1
    ruta = os.path.join(FIGS, f"{_n[0]:02d}_figura.png")
    plt.savefig(ruta, dpi=110, bbox_inches="tight"); plt.close()
    print(f"   [figura guardada: figuras/{os.path.basename(ruta)}]")
plt.show = _save                          # cada plt.show() guarda en vez de abrir

RNG = np.random.default_rng(0)

# ╔═════════════════════════════════════════════════════════════════════════╗
# ║  ✏️  CAMBIA SOLO ESTE BLOQUE POR TU DATASET                             ║
# ╚═════════════════════════════════════════════════════════════════════════╝
DATOS  = os.path.join(AQUI, "..", "..", "..", "data", "ejemplo_regresion.csv")
TARGET = "target"          # nombre EXACTO de la columna respuesta. None si no hay.
# ╚═════════════════════════════════════════════════════════════════════════╝

df = pd.read_csv(DATOS).dropna()

if TARGET is not None:
    y_serie = df[TARGET]; X_df = df.drop(columns=[TARGET])
else:
    y_serie = None; X_df = df.copy()

X_df = pd.get_dummies(X_df, drop_first=True).astype(float)
nombres_X = list(X_df.columns)
X  = X_df.values
Xs = scale(X)

if y_serie is not None:
    y = pd.factorize(y_serie)[0] if y_serie.dtype == object else y_serie.values
    ES_CLASIFICACION = (len(np.unique(y)) <= 10) and np.all(y == np.round(y))
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.30, random_state=0)
    Xs_train, Xs_test, _, _ = train_test_split(Xs, y, test_size=.30, random_state=0)
else:
    y = None; ES_CLASIFICACION = False

def ModeloLineal(**kw):
    return (LogisticRegression(max_iter=5000, **kw) if ES_CLASIFICACION
            else LinearRegression(**kw))

col = y if y is not None else "steelblue"
Z   = PCA().fit_transform(Xs)        # proyección PCA, se usa para graficar
Z2  = Z[:, :2]

print("="*74)
print("A7. Regresión lineal")
print("="*74)
print(f"datos: {X.shape[0]} obs. x {X.shape[1]} predictores")
if y is not None:
    print(f"y ('{TARGET}'): " + ("CLASIFICACIÓN" if ES_CLASIFICACION else "REGRESIÓN"))
print()

# ════════════════════════════════════════════════════════════════════════
#  CÓDIGO DEL MÉTODO
# ════════════════════════════════════════════════════════════════════════

# ═══ A7. Regresión lineal ═══

reg = LinearRegression(   # ⬅ para clasificación usa LogisticRegression
    
    fit_intercept=True,     # ⬅ True casi siempre (incluye el β₀)
).fit(X_train, y_train)

# ── Coeficientes ordenados por magnitud ─────────────────────────────────
coefs = pd.DataFrame({"variable": nombres_X, "coeficiente": reg.coef_})
coefs["|coef|"] = coefs["coeficiente"].abs()
print(f"Intercepto (β₀): {reg.intercept_:.4f}\n")
print(coefs.sort_values("|coef|", ascending=False).to_string(index=False))
print("\nLectura: si la variable sube 1 unidad, y cambia en 'coeficiente'")
print("unidades, MANTENIENDO LAS DEMÁS CONSTANTES.")

pred_lm = reg.predict(X_test)
mse_lm = mean_squared_error(y_test, pred_lm)
print(f"\nMSE test: {mse_lm:.3f}   |   R² test: {r2_score(y_test, pred_lm):.4f}")

# ── Gráfica de coeficientes ─────────────────────────────────────────────
c = coefs.sort_values("coeficiente")
plt.barh(c["variable"], c["coeficiente"], color=np.where(c["coeficiente"]>0, "steelblue", "indianred"))
plt.axvline(0, color="k", lw=.8)
plt.xlabel("coeficiente"); plt.title("Coeficientes de la regresión lineal")
plt.tight_layout(); plt.show()

# ── Tabla completa con valores p (statsmodels) ──────────────────────────
import statsmodels.api as sm
ols = sm.OLS(y_train, sm.add_constant(X_train)).fit()
print(ols.summary().tables[1])
print("\nP>|t| < 0.05  ->  la variable es estadísticamente significativa.")

# ── Diagnóstico de multicolinealidad (VIF) ─────────────────────────────
from statsmodels.stats.outliers_influence import variance_inflation_factor
Xc = sm.add_constant(X_train)
vif = pd.DataFrame({
    "variable": ["const"] + nombres_X,
    "VIF": [variance_inflation_factor(Xc, i) for i in range(Xc.shape[1])]
})
print("\n", vif[vif.variable != "const"].sort_values("VIF", ascending=False).to_string(index=False))
print("VIF > 10  ->  multicolinealidad seria. Usa Ridge/Lasso (tema A9).")

print("\nListo. Revisa la carpeta figuras/")

File: ml-parcial-IIND3118/ml-parcial-IIND3118/temas/A07_regresion_lineal/R/A07_regresion_lineal.R
# ############################################################################
# A7. REGRESIÓN LINEAL
# Tema A07_regresion_lineal — Introducción al Machine Learning (IIND3118)
#
# CÓMO USARLO
#   1. Cambia el bloque marcado con ✏️ por tu dataset.
#   2. Corre:  Rscript A07_regresion_lineal.R     (o ábrelo en RStudio)
#   3. Las gráficas se guardan en la carpeta figuras/ de este tema.
#
# La explicación del método está en el README.md de esta misma carpeta.
# ############################################################################

for (p in c("MASS","cluster","dbscan","glmnet","leaps","caret","class","fpc",
            "clusterCrit","isotree","kernlab","Rtsne","psych","car","vegan")) {
  if (!requireNamespace(p, quietly=TRUE))
    try(install.packages(p, repos="https://cloud.r-project.org"), silent=TRUE)
}
suppressMessages({ library(MASS); library(cluster) })

set.seed(0)

# Las gráficas se guardan en figuras/ en vez de abrirse en pantalla
dir.create("../figuras", showWarnings=FALSE)
.n <- 0
guardar <- function() {
  .n <<- .n + 1
  dev.copy(png, sprintf("../figuras/%02d_figura.png", .n), width=900, height=650)
  invisible(dev.off())
}

# ╔═══════════════════════════════════════════════════════════════════════╗
# ║  ✏️  CAMBIA SOLO ESTE BLOQUE POR TU DATASET                           ║
# ╚═══════════════════════════════════════════════════════════════════════╝
DATOS  <- "../../../data/ejemplo_regresion.csv"
TARGET <- "target"        # nombre EXACTO de la columna respuesta. NULL si no hay.
# ╚═══════════════════════════════════════════════════════════════════════╝

df <- na.omit(read.csv(DATOS))

if (!is.null(TARGET)) {
  y_raw <- df[[TARGET]]
  X_df  <- df[, setdiff(names(df), TARGET), drop=FALSE]
} else { y_raw <- NULL; X_df <- df }

X_df <- as.data.frame(model.matrix(~ . - 1, data=X_df))
nombres_X <- names(X_df)
X  <- as.matrix(X_df)
Xs <- scale(X)

if (!is.null(y_raw)) {
  ES_CLASIFICACION <- is.factor(y_raw) || is.character(y_raw) || length(unique(y_raw)) <= 10
  y <- if (ES_CLASIFICACION) as.factor(y_raw) else as.numeric(y_raw)
  idx <- sample(seq_len(nrow(X)), floor(0.70*nrow(X)))
  Xtrain <- X[idx,]; Xtest <- X[-idx,]; ytrain <- y[idx]; ytest <- y[-idx]
  Xs_train <- Xs[idx,]; Xs_test <- Xs[-idx,]
} else { y <- NULL; ES_CLASIFICACION <- FALSE }

pca  <- prcomp(X, center=TRUE, scale.=TRUE)
Z2   <- pca$x[, 1:2]
col_pt <- if (!is.null(y)) as.integer(as.factor(y)) else "steelblue"

cat(strrep("=",74), "\n")
cat("A7. REGRESIÓN LINEAL\n")
cat(strrep("=",74), "\n")
cat(sprintf("datos: %d obs. x %d predictores\n\n", nrow(X), ncol(X)))

# ############################################################################
#  CÓDIGO DEL MÉTODO
# ############################################################################

# ============================================================================
# A7. REGRESIÓN LINEAL
# ----------------------------------------------------------------------------
# QUÉ HACE: busca la recta (o plano) que mejor pasa por los datos,
# minimizando la suma de errores al cuadrado.
#
#   beta = argmin suma( (Y_i - beta'X_i)^2 )
#
# CÓMO SE INTERPRETA (esto siempre lo preguntan)
#   "Si X_j aumenta UNA UNIDAD, el valor esperado de Y cambia en beta_j
#    unidades, MANTENIENDO TODAS LAS DEMÁS VARIABLES CONSTANTES."
#
# ✅ El más interpretable · rápido · da valores p e intervalos
# ❌ Solo relaciones LINEALES · sensible a outliers · sufre con
#    MULTICOLINEALIDAD (variables muy correlacionadas -> coeficientes
#    inestables y enormes) -> eso es justo lo que arreglan Ridge/Lasso
#
# SUPUESTOS: linealidad, errores independientes, varianza constante, normalidad
# ============================================================================

if (!ES_CLASIFICACION) {
  dtr <- data.frame(Xtrain, ytrain=ytrain)
  reg <- lm(ytrain ~ ., data=dtr)
  print(summary(reg))
  cat("\nP>|t| < 0.05 -> la variable es estadísticamente significativa.\n")

  pred_lm <- predict(reg, data.frame(Xtest))
  mse_lm  <- mean((ytest - pred_lm)^2)
  cat(sprintf("MSE test: %.3f\n", mse_lm))

  # GRÁFICA de coeficientes
  co <- coef(reg)[-1]
  barplot(sort(co), horiz=TRUE, las=1, cex.names=.7,
          col=ifelse(sort(co)>0,"steelblue","indianred"),
          main="Coeficientes de la regresión lineal", xlab="coeficiente")
  abline(v=0)

  # MULTICOLINEALIDAD: VIF > 10 es problema serio
  if (requireNamespace("car", quietly=TRUE) && ncol(Xtrain) > 1) {
    v <- car::vif(reg)
    print(sort(v, decreasing=TRUE))
    cat("VIF > 10 -> multicolinealidad seria. Usa Ridge/Lasso (A9).\n")
  }
}

cat("\nListo. Revisa la carpeta figuras/\n")
